include/exadg/acoustic_conservation_equations/postprocessor/sound_energy_calculator.h

```diff
@@ -0,0 +1,263 @@
+
+#include <deal.II/matrix_free/matrix_free.h>
+
+#include <exadg/matrix_free/integrators.h>
+#include <exadg/utilities/create_directories.h>
+#include <exadg/utilities/print_functions.h>
+
+#include <fstream>
+
+namespace ExaDG
+{
+namespace Acoustics
+{
+struct SoundEnergyCalculatorData
+{
+  SoundEnergyCalculatorData()
+    : directory("output/"),
+      filename("sound_energy.csv"),
+      clear_file(true),
+      density(-1.0),
+      speed_of_sound(-1.0)
+  {
+  }
+
+  void
+  print(dealii::ConditionalOStream & pcout)
+  {
+    if(time_control_data.is_active)
+    {
+      pcout << std::endl << "  Calculate sound energy:" << std::endl;
+
+      // only implemented for unsteady problem
+      time_control_data.print(pcout, true /*unsteady*/);
+
+      print_parameter(pcout, "Directory of output files", directory);
+      print_parameter(pcout, "Filename", filename);
+    }
+  }
+
+  TimeControlData time_control_data;
+
+  // directory and filename
+  std::string directory;
+  std::string filename;
+  bool        clear_file;
+
+  double density;
+  double speed_of_sound;
+};
+
+template<int dim, typename Number>
+class SoundEnergyCalculator
+{
+  using This = SoundEnergyCalculator<dim, Number>;
+
+  using BlockVectorType = dealii::LinearAlgebra::distributed::BlockVector<Number>;
+
+  using scalar = dealii::VectorizedArray<Number>;
+  using vector = dealii::Tensor<1, dim, scalar>;
+
+  using CellIntegratorU = CellIntegrator<dim, dim, Number>;
+  using CellIntegratorP = CellIntegrator<dim, 1, Number>;
+
+public:
+  SoundEnergyCalculator(MPI_Comm const & comm)
+    : mpi_comm(comm),
+      clear_files(true),
+      matrix_free(nullptr),
+      dof_index_pressure(0),
+      dof_index_velocity(1),
+      quad_index(0),
+      block_index_pressure(0),
+      block_index_velocity(1),
+      rho(-1.0),
+      c(-1.0),
+      rhocc_inv(-1.0)
+  {
+  }
+
+  void
+  setup(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+        SoundEnergyCalculatorData const &       data_in,
+        unsigned int const                      dof_index_pressure_in,
+        unsigned int const                      dof_index_velocity_in,
+        unsigned int const                      quad_index_in,
+        unsigned int const                      block_index_pressure_in,
+        unsigned int const                      block_index_velocity_in)
+  {
+    time_control.setup(data_in.time_control_data);
+
+    matrix_free = &matrix_free_in;
+    data        = data_in;
+
+    dof_index_pressure   = dof_index_pressure_in;
+    dof_index_velocity   = dof_index_velocity_in;
+    quad_index           = quad_index_in;
+    block_index_pressure = block_index_pressure_in;
+    block_index_velocity = block_index_velocity_in;
+
+    rho       = static_cast<Number>(data_in.density);
+    c         = static_cast<Number>(data_in.speed_of_sound);
+    rhocc_inv = static_cast<Number>(
+      1.0 / (data_in.density * data_in.speed_of_sound * data_in.speed_of_sound));
```

Comment:
I think this should be in the `cell_loop()` in `calculate_sound_energy()`

Reply:
[2f33618](https://github.com/exadg/exadg/pull/626/commits/2f336186a4229565ad0cccf9a46851d3e668fea6)